cirq/experiments/n_qubit_tomography.py

```diff
@@ -0,0 +1,194 @@
+"""Tomography code for an arbitrary number of qubits allowing for
+different pre-measurement rotations.
+The code is designed to be modular with regards to data collection
+so that occurs outside of the StateTomographyExperiment class.
+"""
+from typing import List, Tuple, Optional
+
+import numpy as np
+import sympy
+
+import cirq
+from cirq.experiments.qubit_characterizations import TomographyResult
+
+
+class StateTomographyExperiment:
+    """ Experiment to conduct state tomography.
+
+    Generates data collection protocol for the state tomography experiment.
+    Does the fitting of generated data to determine the density matrix.
+    Attributes:
+        rot_moment: Moment with parametrized rotation gates to do before the
+            final measurements.
+        rot_sweep: The list of rotations on the qubits to perform before
+            measurement.
+        mat: Matrix of coefficients for the system.  Each row is one equation
+            corresponding to a rotation sequence and bit string outcome for
+            that rotation sequence.  Each column corresponds to the coefficient
+            on one term in the density matrix.
+        num_qubits: Number of qubits to do tomography on.
+    """
+
+    def __init__(self,
+                 qubits: List[cirq.Qid],
+                 prerotations: Optional[List[Tuple[float, float]]] = None):
+        """ Initializes the rotation protocol and matrix for system.
+
+        Args:
+            qubits: qubits to do the tomography on.
+            prerotations: Tuples of (phase_exponent, exponent) parameters for
+                gates to apply to the qubits before measurement. The actual
+                rotation applied will be
+                    cirq.PhasedXPowGate(
+                        phase_exponent=phase_exponent, exponent=exponent)
+                If none, we use [(0, 0), (0, 0.5), (0.5, 0.5)], which
+                corresponds to rotation gates [I, X**0.5, Y**0.5]
+        """
+        if prerotations == None:
+            prerotations = [(0, 0), (0, 0.5), (0.5, 0.5)]
+
+        self.num_qubits = len(qubits)
+
+        phase_exp_vals, exp_vals = zip(*prerotations)  # type: ignore
+
+        ops: List[cirq.Operation] = []
+        sweeps: List[cirq.Sweep] = []
+        for i, qubit in enumerate(qubits):
+            phase_exp = sympy.Symbol(f'phase_exp_{i}')
+            exp = sympy.Symbol(f'exp_{i}')
+            gate = cirq.PhasedXPowGate(phase_exponent=phase_exp, exponent=exp)
+            ops.append(gate.on(qubit))
+            sweeps.append(
+                cirq.Points(phase_exp, phase_exp_vals) +
+                cirq.Points(exp, exp_vals))
+
+        self.rot_moment = cirq.Circuit.from_ops(ops)
+        self.rot_sweep = cirq.Product(*sweeps)
+        self.mat = self.make_state_tomography_matrix()
```

Comment:
I'm not sure what purpose this class is fulfilling.  I might move this init code into `state_tomography` and make `state_tomography` and `get_state_tomography_data` methods of the class.

Reply:
We wanted a class to hold the matrix making code since it can be very expensive.  This allows us to make once class for the tomography and then use it for different data.

Reply:
In that case, `make_state_tomography_matrix` should be private (`_make_state_tomography_matrix`) and the class name and docstring should be changed to indicate the intended usage (`StateTomographyMatrix`?).

Reply:
It is intended to be a wrapper for all the experiment logic including fitting using the matrix.  This way all people have to do is supply the data.

Reply:
If a user calls `make_state_tomography_matrix()`, it just recalculates the value already in the `mat` attribute.

Reply:
Oh yeah for sure I agree it should be private I just meant in terms of naming.